contracts/src/token/erc1155/mod.rs

```diff
@@ -0,0 +1,1812 @@
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `operator` - Account to add to the set of authorized operators.
+    /// * `from` - Account of the sender.
+    /// * `to` - Account of the recipient.
+    /// * `token_ids` - Array of all token id.
+    /// * `values` - Array of all amount of tokens to be transferred.
+    /// * `data` - Additional data with no specified format, sent in call to
+    ///   `to`.
+    ///
+    /// # Errors
+    ///
+    /// If [`IERC1155Receiver::on_erc_1155_received`] hasn't returned its
+    /// interface id or returned with error, then the error
+    /// [`Error::InvalidReceiver`] is returned.
+    fn _check_on_erc1155_batch_received(
+        &mut self,
+        operator: Address,
+        from: Address,
+        to: Address,
+        token_ids: Vec<U256>,
+        values: Vec<U256>,
+        data: Bytes,
+    ) -> Result<(), Error> {
+        const RECEIVER_FN_SELECTOR: FixedBytes<4> = fixed_bytes!("bc197c81");
+
+        if !to.has_code() {
+            return Ok(());
+        }
+
+        let receiver = IERC1155Receiver::new(to);
+        let call = Call::new_in(self);
+        let result = receiver.on_erc_1155_batch_received(
+            call,
+            operator,
+            from,
+            token_ids,
+            values,
+            data.to_vec().into(),
+        );
+
+        let id = match result {
+            Ok(id) => id,
+            Err(e) => {
+                if let call::Error::Revert(ref reason) = e {
+                    if reason.len() > 0 {
+                        // Non-IERC1155Receiver implementer.
+                        return Err(Error::InvalidReceiverWithReason(e));
+                    }
+                }
+
+                return Err(ERC1155InvalidReceiver { receiver: to }.into());
+            }
+        };
+
+        // Token rejected.
+        if id != RECEIVER_FN_SELECTOR {
+            return Err(ERC1155InvalidReceiver { receiver: to }.into());
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::{address, uint, Address, U256};
+    use stylus_sdk::msg;
+
+    use super::{
+        ERC1155InsufficientBalance, ERC1155InvalidArrayLength,
+        ERC1155InvalidOperator, ERC1155InvalidReceiver, ERC1155InvalidSender,
+        ERC1155MissingApprovalForAll, Erc1155, Error, IErc1155,
+    };
+    use crate::{
+        token::erc721::IErc721, utils::introspection::erc165::IErc165,
+    };
+
+    const ALICE: Address = address!("A11CEacF9aa32246d767FCCD72e02d6bCbcC375d");
+    const BOB: Address = address!("F4EaCDAbEf3c8f1EdE91b6f2A6840bc2E4DD3526");
+    const DAVE: Address = address!("0BB78F7e7132d1651B4Fd884B7624394e92156F1");
+    const CHARLIE: Address =
+        address!("B0B0cB49ec2e96DF5F5fFB081acaE66A2cBBc2e2");
+
+    pub(crate) fn random_token_ids(size: usize) -> Vec<U256> {
+        (0..size).map(|_| U256::from(rand::random::<u32>())).collect()
```

Comment:
Do we want random as here? Why not having a sequence? What if we get a collision?